scripts/generate-federate-match.sh

```diff
@@ -0,0 +1,91 @@
+! [ -x "$(command -v uniq)" ] && echo 'uniq not installed, the script requires it.' && exit 1
+! [ -x "$(command -v yq)" ] && echo 'yq not installed, the script requires it.' && exit 1
+
+function log() {
+    echo "${1:-}" >&2
+}
+
+function log_exit() {
+    log "${1:-}"
+
+    exit 1
+}
+
+function get_rules_metrics() {
+    local os_prom_rule_file="${1:-}"
+    [[ "${os_prom_rule_file}" = "" ]] && log_exit "Variable 'os_prom_rule_file' is empty."
+
+    local metrics_list_file="${2:-}"
+    [[ "${metrics_list_file}" = "" ]] && log_exit "Variable 'metrics_list_file' is empty."
+
+    local rules_file
+    rules_file=$(mktemp)
+
+    local json_file
+    json_file=$(mktemp)
+
+    yq '.spec' "${os_prom_rule_file}" > "${rules_file}"
+    mimirtool analyze rule-file "${rules_file}" --output="${json_file}"
+    jq '.ruleGroups[].metrics | select( . != null ) | .[]' "${json_file}" --raw-output >> "${metrics_list_file}"
+
+    rm -rf "${rules_file}"
+    rm -rf "${json_file}"
+}
+
+function append_consolidated_dashboard_fields() {
+    local federation_config_file="${1:-}"
+    [[ "${federation_config_file}" = "" ]] && log_exit "Variable 'federation_config_file' is empty."
+
+    yq --inplace '."match[]" += [ "container_memory_usage_bytes" ]' "${federation_config_file}"
+}
+
+function main() {
+    local script_dir
+    script_dir="$(cd -- "$(dirname -- "${BASH_SOURCE[0]:-$0}")" &>/dev/null && pwd 2>/dev/null)"
+
+    local repo_dir
+    repo_dir=$(realpath "${script_dir}/../")
+
+    local working_tmp_dir
+    working_tmp_dir=$(mktemp -d)
+    log "Created temp dir for storing temp file: '${working_tmp_dir}'"
+
+    local metrics_list_file="${working_tmp_dir}/metrics_list"
+
+    # Get metrics used in Dashboards
+    mimirtool analyze dashboard "${repo_dir}"/resources/grafana/sources/* --output="${working_tmp_dir}/acs.json"
+    jq '.dashboards[].metrics[]' "${working_tmp_dir}/acs.json" --raw-output >> "${metrics_list_file}"
+
+    mimirtool analyze dashboard "${repo_dir}"/resources/mixins/kubernetes/generated/dashboards/* --output="${working_tmp_dir}/mixins.json"
+    jq '.dashboards[].metrics[]' "${working_tmp_dir}/mixins.json" --raw-output >> "${metrics_list_file}"
+
+    # Get metrics used in recording rules and alerts
+    get_rules_metrics "${repo_dir}/resources/prometheus/billing-rules.yaml" "${metrics_list_file}"
+    get_rules_metrics "${repo_dir}/resources/prometheus/kubernetes-mixin-alerts.yaml" "${metrics_list_file}"
+    get_rules_metrics "${repo_dir}/resources/prometheus/prometheus-rules.yaml" "${metrics_list_file}"
+    get_rules_metrics "${repo_dir}/resources/prometheus/rhacs-recording-rules.yaml" "${metrics_list_file}"
+
+    # Filter metrics (exclude metrics that are collected by observability Prometheus or created by recording rules)
+    sort "${metrics_list_file}" | uniq | grep -v -E "^acs|^rox|^aws|^central:|acscs_worker_nodes" > "${metrics_list_file}.filter"
```

Comment:
I don't have any experience in this area, but how can we make sure that the assertion in the comment always stands for example for metrics whose name starts with `aws`?

Reply:
I'm not sure if we can ensure with any "static" checks that some specific metrics are scraped by `rhacs-observability` Prometheus. We know what target we are scraping that is defined in `pod_monitors` (an example: `resources/prometheus/pod_monitors/rhacs-cloudwatch-exporter.yaml`) - but what metric names will be returned by that endpoint, we don't know.

One option is to leave metrics in the list of federated metrics. If they exist in OSD Prometheus, they will be federated; if not, they should not make any impact.